frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/outline/outline_block_component.dart

```diff
@@ -0,0 +1,217 @@
+    with BlockComponentConfigurable {
+  @override
+  BlockComponentConfiguration get configuration => widget.configuration;
+
+  @override
+  Node get node => widget.node;
+
+  late EditorState editorState = context.read<EditorState>();
+  late Stream<(TransactionTime, Transaction)> stream =
+      editorState.transactionStream;
+
+  @override
+  Widget build(BuildContext context) {
+    return StreamBuilder(
+      stream: stream,
+      builder: (context, snapshot) {
+        if (widget.showActions && widget.actionBuilder != null) {
+          return BlockComponentActionWrapper(
+            node: widget.node,
+            actionBuilder: widget.actionBuilder!,
+            child: _buildOutlineBlock(),
+          );
+        }
+        return _buildOutlineBlock();
+      },
+    );
+  }
+
+  Widget _buildOutlineBlock() {
+    return Container(
+      padding: const EdgeInsets.symmetric(
+        horizontal: 15.0,
+        vertical: 20.0,
+      ),
+      decoration: BoxDecoration(
+        border: Border.all(color: Colors.white54),
+        borderRadius: BorderRadius.circular(15.0),
+      ),
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          Text(
+            LocaleKeys.document_plugins_outline_heading.tr(),
+            style: editorState.editorStyle.textStyleConfiguration.text,
+          ),
+          const Divider(
+            color: Colors.white54,
+          ),
+          ...getHeadingNodes().map((e) => OutlineItemWidget(node: e)).toList(),
+        ],
+      ),
+    );
+  }
+
+  Iterable<Node> getHeadingNodes() {
+    final children = editorState.document.root.children;
+    return children.where((element) => element.type == HeadingBlockKeys.type);
+  }
+}
+
+class OutlineItemWidget extends StatelessWidget {
+  OutlineItemWidget({
+    super.key,
+    required this.node,
+  }) {
+    assert(node.type == HeadingBlockKeys.type);
+  }
+
+  final Node node;
+
+  @override
+  Widget build(BuildContext context) {
+    final editorState = context.read<EditorState>();
+    return MouseRegion(
+      cursor: SystemMouseCursors.click,
+      child: GestureDetector(
+        onTap: () {
+          // when clicked scroll the view to the heading
+          editorState.updateSelectionWithReason(
+            Selection.single(
+              path: node.path,
+              startOffset: node.delta?.length ?? 0,
+            ),
+            reason: SelectionUpdateReason.uiEvent,
+          );
+        },
+        child: Container(
+          margin: EdgeInsets.only(top: node.verticalSpacing),
+          padding: EdgeInsets.only(left: node.leftIndent),
+          child: Text(
+            node.outlineItemText,
+            style: editorState.editorStyle.textStyleConfiguration.text,
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+extension on Node {
+  double get verticalSpacing {
+    if (type != HeadingBlockKeys.type) {
+      assert(false);
```

Comment:
Just noticed this assert, feels like a hack to me.

Reply:
because only node with heading type can fetch this values.

Reply:
Yes, but `assert(false)` is not meaningful, and does not provide any information that would lead one to resolve it.

Reply:
Agreed.